x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -809,7 +824,7 @@ private static void throwAlreadyEvicted() {
         boolean tryRead(ByteBuffer buf, long offset) throws IOException {
             int startingPos = buf.position();
             sharedBytes.getFileChannel(sharedBytesPos).read(buf, physicalStartOffset() + getRegionRelativePosition(offset));
-            if (isEvicted() || hasReferences() == false) {
```

Comment:
For my own understanding: we don't need to check `hasReferences()` here because incrementing the reference count automatically sets the `evicted` field?

Reply:
It's more that we cannot have a ref count of `0` unless `evicted` is set, so there really isn't much of a point in checking the ref count.
So ref count == 0 -> evicted, making that check redundant (and costly in the common case of "not evicted")

Reply:
That said, we could technically only check the ref count here instead of the evicted flag. My thinking around this was that the evicted flag is less contended than that count in cash trashing scenarios so it might behave better CPU cache wise. The upside of checking the ref count is, that technically for half-written pages, you may get a valid read in on an evicted page (evicted already set but ref count > 0 still because there's still a writer open for it and thus it's not on the free list).
I decided to go with the flag for now. I have some ideas for speeding this up one more step anyway that changes this a little more and hopefully makes the ref count go away.